k2/csrc/array.h

```diff
@@ -1,69 +1,133 @@
+//                      Fangjun Kuang (csukuangfj@gmail.com)
 
 // See ../../LICENSE for clarification regarding multiple authors
 
 #ifndef K2_CSRC_ARRAY_H_
 #define K2_CSRC_ARRAY_H_
 
-#include <functional>
-#include <limits>
-#include <memory>
+#include <type_traits>
 #include <vector>
 
+#include "glog/logging.h"
+
 namespace k2 {
 
 /*
-   We will use e.g. StridedPtr<int32_t, T> when the stride is not 1, and
+   We will use e.g. StridedPtr<T*, I> when the stride is not 1, and
    otherwise just T* (which presumably be faster).
 */
-template <typename I, typename T>
+template <typename Ptr, typename I = int32_t>
 struct StridedPtr {
-  T *data;
-  I stride;
-  T &operator[](I i) { return data[i]; }
+  using T = typename std::remove_pointer<Ptr>::type;
+
+  T *data;   // it is NOT owned here
+  I stride;  // in number of elements, NOT number of bytes.
+
+  operator T *() { return data; }
+  operator const T *() const { return data; }
+
+  // ++it
+  StridedPtr &operator++() {
+    data += stride;
+    return *this;
+  }
+
+  // it++
+  StridedPtr operator++(int) {
+    StridedPtr tmp(*this);
+    data += stride;
+    return tmp;
+  }
+
+  bool operator==(const StridedPtr &other) const { return data == other.data; }
+
   StridedPtr(T *data, I stride) : data(data), stride(stride) {}
 };
 
-/* MIGHT NOT NEED THIS */
-template <typename I, typename Ptr>
-struct Array1 {
-  // Irregular two dimensional array of something, like vector<vector<X> >
+template <typename Ptr, typename I = int32_t>
+class Array1 {
+ public:
+  // Irregular two dimensional array of something, like vector<X>
   // where Ptr is, or behaves like, X*.
   using IndexT = I;
   using PtrT = Ptr;
+  using ValueType = typename std::remove_pointer<PtrT>::type;
+  using PtrWithStride = StridedPtr<Ptr, I>;
+
+  Array1() : begin_(0), end_(0), stride_(1), data_(nullptr) {}
+
+  Array1(IndexT begin, IndexT end, IndexT stride, PtrT data)
+      : begin_(begin), end_(end), stride_(stride), data_(data) {}
+
+  void Init(IndexT begin, IndexT end, IndexT stride, PtrT data) {
+    begin_ = begin;
+    end_ = end;
+    stride_ = stride;
+    data_ = data;
+  }
+
+  PtrWithStride begin() const {
+    PtrWithStride res(data_ + begin_ * stride_, stride_);
+    return res;
+  }
+
+  PtrWithStride end() const {
+    PtrWithStride res(data_ + end_ * stride_, stride_);
+    return res;
+  }
+
+  bool Empty() const { return begin_ == end_ || data_ == nullptr; }
+
+  IndexT Size() const { return end_ - begin_; }
+
+  ValueType &operator[](IndexT i) {
+    DCHECK_GE(i, 0);
+    DCHECK_LT(i, Size());
+    return data_[begin_ + i * stride_];
+  }
 
+  const ValueType &operator[](IndexT i) const {
+    DCHECK_GE(i, 0);
+    DCHECK_LT(i, Size());
+    return data_[begin_ + i * stride_];
+  }
+
+ private:
   // 'begin' and 'end' are the first and one-past-the-last indexes into `data`
   // that we are allowed to use.
-  IndexT begin;
-  IndexT end;
+  IndexT begin_;
+  IndexT end_;
+  IndexT stride_;
```

Comment:
Incidentally, you shouldn't have added this stride_ member without asking me!  Lots of time wasted.
Also, this object was a struct for a reason.  It is mostly intended as a data structure that outside code will
manipulate.
But focus on Array2 for now.  Array1 is likely not needed.